### PR TITLE
cache parsed line result

### LINKED ISSUE #620: Cache clean string of a line
Currently `RstParser::clean($line)` is being called very often with some costly `substr`'s. What would you think about caching the result of a clean line? This could be done with a class wrapping /  representing a line:

```php
final class Line
{
    private string $raw;
    private string $clean;
    private bool $blank;

    public function __construct(string $line)
    {
        $this->raw = $line;
        $this->clean = RstParser::clean($line);
        $this->blank = '' === $this->clean;
    }

    public function clean(): string
    {
        return $this->clean;
    }

    public function raw(): string
    {
        return $this->raw;
    }

    public function isBlank(): bool
    {
        return $this->blank;
    }
}
```

### PR DESCRIPTION
**Pull request by smoench**
resolves #620 
Sorry, could not resist to refactor things 🙈

### CODE REVIEW

File: src/Rule/BlankLineAfterFilepathInYamlCodeBlock.php [hunk 1]
```diff
@@ -25,8 +25,6 @@ class BlankLineAfterFilepathInYamlCodeBlock extends AbstractRule implements Rule
 {
     public function check(Lines $lines, int $number): ?string
     {
-        $lines = $lines->toIterator();
```
Comment (OskarStark) [inline src/Rule/BlankLineAfterFilepathInYamlCodeBlock.php]: That was on my todo list as well 😊 it was just a “Zwischenschritt” in the past 😁

File: src/Traits/DirectiveTrait.php [hunk 2]
```diff
@@ -13,32 +13,32 @@
...
-use App\Helper\Helper;
 use App\Rst\RstParser;
+use App\Value\Lines;
 
 trait DirectiveTrait
 {
-    private function in(string $directive, \ArrayIterator $lines, int $number, array $directiveTypes = null): bool
+    private function in(string $directive, Lines $lines, int $number, array $directiveTypes = null): bool
     {
-        $lines = Helper::cloneIterator($lines, $number);
```
Comment (OskarStark) [inline src/Traits/DirectiveTrait.php]: I did this because otherwise the injected iterator does not get back to its initial position, does this work now?

Comment (smoench) [reply src/Traits/DirectiveTrait.php]: Well, in the `AnalyseCommand` there is still a separate Iterator being used. The Rules are using the `seek`-method so they can jump back themselves.

**Comment by OskarStark** [comment]
All tests passing? 🤔

Simon Mönch <notifications@github.com> schrieb am Do. 20. Feb. 2020 um
09:15:

**Comment by smoench** [comment]
Yes, but we should add a phpunit github action workflow

**Comment by OskarStark** [comment]
> Yes, but we should add a phpunit github action workflow

We should do this before merging this PR

**Comment by OskarStark** [comment]
Done, can you rebase?

**Pull request merged**